arrow-udf-js/src/jsarrow.rs

```diff
@@ -152,116 +110,422 @@ macro_rules! build_array {
     }};
 }
 
-/// Build arrow array from JS objects.
-pub fn build_array<'a>(field: &Field, ctx: &Ctx<'a>, values: Vec<Value<'a>>) -> Result<ArrayRef> {
-    match field.data_type() {
-        DataType::Null => build_array!(NullBuilder, ctx, values),
-        DataType::Boolean => build_array!(BooleanBuilder, ctx, values),
-        DataType::Int8 => build_array!(Int8Builder, ctx, values),
-        DataType::Int16 => build_array!(Int16Builder, ctx, values),
-        DataType::Int32 => build_array!(Int32Builder, ctx, values),
-        DataType::Int64 => build_array!(Int64Builder, ctx, values),
-        DataType::UInt8 => build_array!(UInt8Builder, ctx, values),
-        DataType::UInt16 => build_array!(UInt16Builder, ctx, values),
-        DataType::UInt32 => build_array!(UInt32Builder, ctx, values),
-        DataType::UInt64 => build_array!(UInt64Builder, ctx, values),
-        DataType::Float32 => build_array!(Float32Builder, ctx, values),
-        DataType::Float64 => build_array!(Float64Builder, ctx, values),
-        DataType::Utf8 => match field
-            .metadata()
-            .get("ARROW:extension:name")
-            .map(|s| s.as_str())
-        {
-            Some("arrowudf.json") => {
-                let mut builder = StringBuilder::with_capacity(values.len(), 1024);
+macro_rules! build_json_array {
+    ($array_type: ty, $ctx:expr, $values:expr) => {{
+        let mut builder = <$array_type>::with_capacity($values.len(), 1024);
+        for val in $values {
+            if val.is_null() || val.is_undefined() {
+                builder.append_null();
+            } else if let Some(s) = $ctx.json_stringify(val)? {
+                builder.append_value(s.to_string()?);
+            } else {
+                builder.append_null();
+            }
+        }
+        Ok(Arc::new(builder.finish()))
+    }};
+}
+
+#[derive(Debug, Clone)]
+pub struct Converter {
+    arrow_extension_key: String,
+    json_extension_name: String,
+    decimal_extension_name: String,
+}
+
+impl Converter {
+    pub fn new() -> Self {
+        Self {
+            arrow_extension_key: "ARROW:extension:name".to_string(),
```

Comment:
I'm not sure if this should be made configurable, as it is defined in Arrow specifications. Any extension types should use this key exactly, otherwise they can not be recognized as extension type in pyarrow or arrow java sdk.

Reply:
Hmm, yeah I see what you're saying. My only thought was that it does improve the flexibility of the library for cases where you have a record batch with columns that you know are json and you want to pass the record batch to a udf without further modification. If I take this change out I'll just need to figure a somewhat ergonomic way to modify the field metadata of an already constructed batch for my use case

Reply:
I see. Let's keep it for flexibility.